src/test/java/de/hysky/skyblocker/utils/InstancedUtilsTest.java

```diff
@@ -0,0 +1,93 @@
+package de.hysky.skyblocker.utils;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public class InstancedUtilsTest {
+
+	@Test
+	void testSameInstanceEqual() {
+		Vector3i vec1 = new Vector3i(8, 8, 8);
+		
+		Assertions.assertEquals(vec1, vec1);
+	}
+	
+	@Test
+	void testSameFieldValuesEqual() {
+		Vector3i vec1 = new Vector3i(8, 8, 8);
+		Vector3i vec2 = new Vector3i(8, 8, 8);
+		
+		Assertions.assertEquals(vec1, vec2);
+	}
+	
+	@Test
+	void testDifferentFieldValuesEqual() {
+		Vector3i vec1 = new Vector3i(8, 8, 8);
+		Vector3i vec2 = new Vector3i(-8, -8, -8);
+		
+		Assertions.assertNotEquals(vec1, vec2);
+	}
+	
+	@Test
+	void testHashCodeOfEqualFieldValues() {
+		Vector3i vec1 = new Vector3i(8, 8, 8);
+		Vector3i vec2 = new Vector3i(8, 8, 8);
+		
+		Assertions.assertEquals(vec1.hashCode(), vec2.hashCode());
+	}
+	
+	@Test
+	void testHashCodeOfDifferentFieldValues() {
+		Vector3i vec1 = new Vector3i(8, 8, 8);
+		Vector3i vec2 = new Vector3i(-8, -8, -8);
+		
+		Assertions.assertNotEquals(vec1.hashCode(), vec2.hashCode());
+	}
+	
+	@Test
+	void testToString() {
+		Vector3i vec1 = new Vector3i(8, 8, 8);
+		
+		Assertions.assertEquals(vec1.toString(), "Vector3i[x=8, y=8, z=8]");
+	}
+	
+	@SuppressWarnings("unused")
+	private static class Vector3i {
+		final int x;
+		final int y;
+		final int z;
+		
+		Vector3i(int x, int y, int z) {
+			this.x = x;
+			this.y = y;
+			this.z = z;
+		}
+		
+		@Override
+		public boolean equals(Object o) {
+			try {
+				return (boolean) InstancedUtils.equals(getClass()).invokeExact(this, o);
+			} catch (Throwable ignored) {
+				return false;
```

Comment:
Maybe we should throw or return default equals?

Reply:
Perhaps the default equals - either way this should never throw an exception just that the compiler doesn’t allow you to ignore it.